stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala

```diff
@@ -428,6 +428,57 @@ object Flow {
    */
   def fromFunction[A, B](f: A => B): Flow[A, B, NotUsed] = apply[A].map(f)
 
+  /**
+   * Creates a FlowW from an existing base Flow outputting an optional element and
+   * applying an additional viaFlow only if the element in the stream is defined.
+   *
+   * '''Emits when''' the provided viaFlow is runs with defined elements
+   *
+   * '''Backpressures when''' the viaFlow runs for the defined elements and downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @param flow The base flow that outputs an optional element
+   * @param viaFlow The flow that gets used if the optional element in is defined.
+   * @param combine How to combine the materialized values of flow and viaFlow
+   * @return a Flow with the viaFlow applied onto defined elements of the flow. The output value
+   *         is contained within an Option which indicates whether the original flow's element had viaFlow
+   *         applied.
+   * @since 1.1.0
+   */
+  def optionalVia[FIn, FOut, FViaOut, FMat, FViaMat, Mat](flow: Flow[FIn, Option[FOut], FMat],
```

Comment:
It seems like we can just change it to something like `predicate: Out =>Boolen` for conditional transformation, instead of been forcing to lift to an `Optional/Option`, which can introduce additional allocation.

Reply:
+1

Reply:
The optional is necessary for the problem I am solving since

1. I need to keep the elements that don't fulfill the predicate (i.e. I am using `StreamWithContext`/`FlowWithContext` and I don't want to lose the context)
2. I want to know which elements were filtered and which weren't after the matter of fact, so even if this was changed to a predicate it wouldn't ultimately save on performance as I would manually have to add the functionality to achieve this.

I am not against adding an alternative version that avoids boxing due to using a predicate but I would suggest to do that in a different PR, and as @raboof pointed out earlier that may not even need a specific utility function depending on whether point 2 is important or not.

Reply:
The optional is just a kind of marker field, you can have an object with a field `additionalProcessEnabled:Boolean` for that case.

I suggest we look at fs2/zio/reactor-core for the same case before we merge this.

Reply:
> The optional is just a kind of marker field, you can have an object with a field `additionalProcessEnabled:Boolean` for that case.
But in that case the types would then be different which means it would be a different API/overloaded function etc etc (which I don't have a problem with).

> I suggest we look at fs2/zio/reactor-core for the same case before we merge this.

Sure, not that familiar with these API's but I can have a look at them a bit later